src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py

```diff
@@ -43,9 +43,6 @@ def __init__(self, handler: Any) -> None:
         self.queue: List[ImageSpec] = []
         self.ioctxs: Dict[Tuple[str, str], Tuple[rados.Ioctx, Set[ImageSpec]]] = {}
 
-    def __del__(self) -> None:
-        self.wait_for_pending()
```

Comment:
nit: I would prefer to have `assert not self.pending` then here so we could catch a case when `wait_for_pending` was not called.

Reply:
Discussed with @idryomov  regarding whether the \_\_del\_\_() method is needed for `CreateSnapshotRequests` class.

Since  https://github.com/ceph/ceph/pull/32051/commits/3363a1001c7bb95763cdb7d8a9c2ebaff35a2cdd , the ceph-mgr on receiving SIGINT or SIGTERM doesn't request the python modules to cleanly shutdown and garbage collect its resources. It relies on the ceph-mon to blocklist the mgr modules's clients and exits.  I don't see how `Mgr::handle_signal()`  https://github.com/ceph/ceph/blob/v18.2.0/src/mgr/Mgr.cc#L217 that calls `Mgr::shutdown()` gets called anymore. If Mgr::shutdown() doesn't get called, I also don't see how the python MgrModule.shutdown() gets called.

I see couple of python mgr modules, e.g., (FS) volumes module, having a  __del__() method. In the case of the volumes module, the `shutdown()` and `__del__()` were introduced by https://github.com/ceph/ceph/commit/2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174  as part of the PR https://github.com/ceph/ceph/pull/30890 that expected Mgr::shutdown() to be called that wasn't the case after https://github.com/ceph/ceph/pull/32051/

In rbd_support module, the module explicitly calls the `shutdown()` of the handlers and the parent module during recovery from blocklisting in the module's recovery thread, https://github.com/ceph/ceph/blob/v18.2.0/src/pybind/mgr/rbd_support/module.py#L123. I don't think the rbd_support module's shutdown() can be called outside the recovery thread.

Reply:
> Discussed with @idryomov regarding whether the __del__() method is needed for `CreateSnapshotRequests` class.
> 
> Since [3363a10](https://github.com/ceph/ceph/commit/3363a1001c7bb95763cdb7d8a9c2ebaff35a2cdd) , the ceph-mgr on receiving SIGINT or SIGTERM doesn't request the python modules to cleanly shutdown and garbage collect its resources. It relies on the ceph-mon to blocklist the mgr modules's clients and exits. I don't see how `Mgr::handle_signal()` https://github.com/ceph/ceph/blob/v18.2.0/src/mgr/Mgr.cc#L217 that calls `Mgr::shutdown()` gets called anymore.

I would remove these then.  The "simply exit" patch may have been intended to be temporary (like all workarounds are!), but it's been 4 years...

> If Mgr::shutdown() doesn't get called, I also don't see how the python MgrModule.shutdown() gets called.
 and that is quite a lot of dead code...

Python `MgrModule.shutdown()` is called from C++ `PyModuleRunner::shutdown()`, which is called from `ActivePyModules::shutdown()` and `StandbyPyModules::shutdown()`.  It looks like the former is dead code (as you noted above), but the latter appears to be alive -- see `PyModuleRegistry::active_start()`.

> 
> I see couple of python mgr modules, e.g., (FS) volumes module, having a **del**() method. In the case of the volumes module, the `shutdown()` and `__del__()` were introduced by [2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174](https://github.com/ceph/ceph/commit/2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174) as part of the PR #30890 that expected Mgr::shutdown() to be called that wasn't the case after #32051

What I don't understand there is why in #30890 it was deemed necessary to call `VolumeClient.shutdown()` from both `Module.shutdown()` and `Module.__del__()`.  It doesn't make sense to me because:
- `Module.__del__()` would end up doing shutdown work only if the last reference to `Module` somehow got put _without_  `Module.shutdown()` being called first.  Otherwise there would be no shutdown work to do.
- It's the manager framework who was expected to call `Module.shutdown()` from C++ via `PyModuleRunner::shutdown()`.
- It's the (same) manager framework who is holding at least one reference to `Module`.

So either I'm missing something, or all `__del__()` methods in modules (`MgrModule` subclasses) where something is closed or "shut down" are bogus.  `__del__()` isn't even guaranteed to be called:

https://docs.python.org/3.11/reference/datamodel.html

> An implementation is allowed to postpone garbage collection or omit it altogether — it is a matter of implementation quality how garbage collection is implemented [...]

> [...] Do not depend on immediate finalization of objects when they become unreachable (so you should always close files explicitly).

Reply:
And worse:

> If a base class has a `__del__()` method, the derived class’s `__del__()` method, if any, must explicitly call it to ensure proper deletion of the base class part of the instance.

`MgrModule` does have it:
```
class MgrModule(ceph_module.BaseMgrModule, MgrModuleLoggingMixin):
...
    def __del__(self) -> None:
        self._unconfigure_logging()
```
but `volumes` module doesn't do that:
```
class Module(orchestrator.OrchestratorClientMixin, MgrModule):
...
    def __del__(self):
        self.vc.shutdown()
```

Reply:
> > I don't see how `Mgr::handle_signal()` https://github.com/ceph/ceph/blob/v18.2.0/src/mgr/Mgr.cc#L217 that calls `Mgr::shutdown()` gets called anymore.
> 
> I would remove these then. The "simply exit" patch may have been intended to be temporary (like all workarounds are!), but it's been 4 years...
> 

I can do this in a separate PR. I don't think we need to block this PR for that change. What do you think?

> > If Mgr::shutdown() doesn't get called, I also don't see how the python MgrModule.shutdown() gets called.
> 
> Python `MgrModule.shutdown()` is called from C++ `PyModuleRunner::shutdown()`, which is called from `ActivePyModules::shutdown()` and `StandbyPyModules::shutdown()`. It looks like the former is dead code (as you noted above), but the latter appears to be alive -- see `PyModuleRegistry::active_start()`.
> 

Good catch! It looks like only dashboard and prometheus modules have a standby mode (i.e., inherit from MgrStandbyModule). I guess shutdown() methods of only those modules would be called in .  `PyModuleRegistry::active_start()` .

> > I see couple of python mgr modules, e.g., (FS) volumes module, having a **del**() method. In the case of the volumes module, the `shutdown()` and `__del__()` were introduced by [2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174](https://github.com/ceph/ceph/commit/2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174) as part of the PR #30890 that expected Mgr::shutdown() to be called that wasn't the case after #32051
> 
> What I don't understand there is why in #30890 it was deemed necessary to call `VolumeClient.shutdown()` from both `Module.shutdown()` and `Module.__del__()`.  

@vshankar would you know why this was introduced in #30890?

> It doesn't make sense to me because:
> 
> * `Module.__del__()` would end up doing shutdown work only if the last reference to `Module` somehow got put _without_  `Module.shutdown()` being called first.  Otherwise there would be no shutdown work to do.
> * It's the manager framework who was expected to call `Module.shutdown()` from C++ via `PyModuleRunner::shutdown()`.
> * It's the (same) manager framework who is holding at least one reference to `Module`.
> 
> So either I'm missing something, or all `__del__()` methods in modules (`MgrModule` subclasses) where something is closed or "shut down" are bogus. `__del__()` isn't even guaranteed to be called:
> 
> https://docs.python.org/3.11/reference/datamodel.html
> 
> > An implementation is allowed to postpone garbage collection or omit it altogether — it is a matter of implementation quality how garbage collection is implemented [...]
> 
> > [...] Do not depend on immediate finalization of objects when they become unreachable (so you should always close files explicitly).

Thanks for explaining your rationale and sharing the useful reference regarding `__del__()`

Reply:
> And worse:
> 
> > If a base class has a `__del__()` method, the derived class’s `__del__()` method, if any, must explicitly call it to ensure proper deletion of the base class part of the instance.
> 
> `MgrModule` does have it:
> 
> ```
> class MgrModule(ceph_module.BaseMgrModule, MgrModuleLoggingMixin):
> ...
>     def __del__(self) -> None:
>         self._unconfigure_logging()
> ```
> 
> but `volumes` module doesn't do that:
> 
> ```
> class Module(orchestrator.OrchestratorClientMixin, MgrModule):
> ...
>     def __del__(self):
>         self.vc.shutdown()
> ```

Good catch! I think `__del__()`  of  the volumes module can be removed in a separate PR

Reply:
> > I would remove these then. The "simply exit" patch may have been intended to be temporary (like all workarounds are!), but it's been 4 years...
> 
> I can do this in a separate PR. I don't think we need to block this PR for that change. What do you think?

> Good catch! I think `__del__()` of the volumes and the other module can be removed in a separate PR

+1, definitely not in this PR.

Reply:
> > > I don't see how `Mgr::handle_signal()` https://github.com/ceph/ceph/blob/v18.2.0/src/mgr/Mgr.cc#L217 that calls `Mgr::shutdown()` gets called anymore.
> > 
> > 
> > I would remove these then. The "simply exit" patch may have been intended to be temporary (like all workarounds are!), but it's been 4 years...
> 
> I can do this in a separate PR. I don't think we need to block this PR for that change. What do you think?
> 
> > > If Mgr::shutdown() doesn't get called, I also don't see how the python MgrModule.shutdown() gets called.
> > 
> > 
> > Python `MgrModule.shutdown()` is called from C++ `PyModuleRunner::shutdown()`, which is called from `ActivePyModules::shutdown()` and `StandbyPyModules::shutdown()`. It looks like the former is dead code (as you noted above), but the latter appears to be alive -- see `PyModuleRegistry::active_start()`.
> 
> Good catch! It looks like only dashboard and prometheus modules have a standby mode (i.e., inherit from MgrStandbyModule). I guess shutdown() methods of only those modules would be called in . `PyModuleRegistry::active_start()` .
> 
> > > I see couple of python mgr modules, e.g., (FS) volumes module, having a **del**() method. In the case of the volumes module, the `shutdown()` and `__del__()` were introduced by [2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174](https://github.com/ceph/ceph/commit/2eb0c503047b#diff-df2079eaaa6bf726076d8b3e4f8aaa50a9ac1a904e458e8a21247e5bce0702a5R174) as part of the PR #30890 that expected Mgr::shutdown() to be called that wasn't the case after #32051
> > 
> > 
> > What I don't understand there is why in #30890 it was deemed necessary to call `VolumeClient.shutdown()` from both `Module.shutdown()` and `Module.__del__()`.
> 
> @vshankar would you know why this was introduced in #30890?

(apologies for chipping in late)

I don't recall clearly, sorry - had to do something with cleanup during termination and delayed unloading.